internal/storage/blobstorage/gcs/blob_storage.go

```diff
@@ -0,0 +1,227 @@
+	return &blobStorageFactory{params}
+}
+
+// Create implements BlobStorageFactory.
+func (f *blobStorageFactory) Create() (internal.BlobStorage, error) {
+	return New(f.params)
+}
+
+func New(params BlobStorageParams) (internal.BlobStorage, error) {
+	metrics := params.Metrics.SubScope("gcs_blob_storage")
+	if params.Config.GCP == nil {
+		return nil, xerrors.Errorf("GCP project id not configured")
+	}
+	if len(params.Config.GCP.Bucket) == 0 {
+		return nil, xerrors.Errorf("GCP bucket not configure for blob storage")
+	}
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to create GCS client, error: %w", err)
+	}
+	blobStorageMetrics := &blobStorageMetrics{
+		blobDownloadedSize: metrics.SubScope(blobDownloaderScopeName).Timer(blobSizeMetricName),
+		blobUploadedSize:   metrics.SubScope(blobUploaderScopeName).Timer(blobSizeMetricName),
+	}
+	return &blobStorageImpl{
+		logger:             log.WithPackage(params.Logger),
+		config:             params.Config,
+		project:            params.Config.GCP.Project,
+		bucket:             params.Config.GCP.Bucket,
+		client:             client,
+		blobStorageMetrics: blobStorageMetrics,
+		instrumentUpload:   instrument.NewWithResult[string](metrics, "upload"),
+		instrumentDownload: instrument.NewWithResult[*api.Block](metrics, "download"),
+	}, nil
+}
+
+func (s *blobStorageImpl) Upload(ctx context.Context, block *api.Block, compression api.Compression) (string, error) {
+	return s.instrumentUpload.Instrument(ctx, func(ctx context.Context) (string, error) {
+		var key string
+		defer s.logDuration("upload", time.Now())
+
+		// Skip the upload if the block itself is skipped.
+		if block.Metadata.Skipped {
+			return "", nil
+		}
+
+		data, err := proto.Marshal(block)
+		if err != nil {
+			return "", xerrors.Errorf("failed to marshal block: %w", err)
+		}
+
+		blockchainNetwork := fmt.Sprintf("%s/%s", block.Blockchain, block.Network)
+		tagHeightHash := fmt.Sprintf("%d/%d/%s", block.Metadata.Tag, block.Metadata.Height, block.Metadata.Hash)
+		if s.config.Chain.Sidechain != api.SideChain_SIDECHAIN_NONE {
+			key = fmt.Sprintf(
+				"%s/%s/%s", blockchainNetwork, block.SideChain, tagHeightHash,
+			)
+		} else {
+			key = fmt.Sprintf(
+				"%s/%s", blockchainNetwork, tagHeightHash,
+			)
+		}
+
+		data, err = storage_utils.Compress(data, compression)
+		if err != nil {
+			return "", xerrors.Errorf("failed to compress data with type %v: %w", compression.String(), err)
+		}
+		key, err = storage_utils.GetObjectKey(key, compression)
+		if err != nil {
+			return "", xerrors.Errorf("failed to get object key: %w", err)
+		}
+
+		// #nosec G401
+		h := md5.New()
+		size, err := h.Write(data)
+		if err != nil {
+			return "", xerrors.Errorf("failed to compute checksum: %w", err)
+		}
+
+		checksum := h.Sum(nil)
+
+		object := s.client.Bucket(s.bucket).Object(key)
+		w := object.NewWriter(ctx)
+		_, err = w.Write(data)
```

Comment:
Instead of doing the network call to fetch the object.Attributes after the upload, you can just set the expected MD5 before the write.  Something like `w.Attrs().MD5 = checksum`.  Blob storage should fail the upload if the MD5 does not match.

Reply:
But it does not reject the upload as it says in the doc. Am I missing something?

<img width="655" alt="image" src="https://github.com/coinbase/chainstorage/assets/432106/5a4a8439-06b3-4c1f-8590-435172513315">

<img width="900" alt="image" src="https://github.com/coinbase/chainstorage/assets/432106/092856fe-9bb8-4306-9b50-b187abd8f4d2">

Reply:
w.Attrs().MD5 and w.MD5 are 2 different objects, I'm not sure if that makes a difference.

Reply:
![image](https://github.com/coinbase/chainstorage/assets/432106/3077bc79-3c9f-4e0e-a1b3-a84cd96382b6)

Yes, indeed we can avoid an extra request.